mypy/report.py

```diff
@@ -103,6 +104,115 @@ def on_finish(self) -> None:
 reporter_classes['linecount'] = LineCountReporter
 
 
+class LineCoverageVisitor(TraverserVisitor):
+    def __init__(self, source: List[str]) -> None:
+        self.source = source
+
+        # For each line of source, we maintain a pair of
+        #  * the indentation level of the surrounding function
+        #    (-1 if not inside a function), and
+        #  * whether the surrounding function is typed.
+        # Initially, everything is covered at indentation level -1.
+        self.lines_covered = [(-1, True) for l in source]
+
+    # The Python AST has position information for the starts of
+    # elements, but not for their ends. Fortunately the
+    # indentation-based syntax makes it pretty easy to find where a
+    # block ends without doing any real parsing.
+
+    # TODO: Handle line continuations (explicit and implicit) and
+    # multi-line string literals. (But at least line continuations
+    # are normally more indented than their surrounding block anyways,
+    # by PEP 8.)
+
+    def indentation_level(self, line_number: int) -> Optional[int]:
+        """Return the indentation of a line of the source (specified by
+        zero-indexed line number). Returns None for blank lines or comments."""
```

Comment:
I took a quick look at having typed_ast give the ending position, at least for function definitions. It doesn't look that hard, but then I realized I don't even know how to build typed_ast and I don't really want to get into that right now.

Another approach would be to just take the largest (starting) line of a node within a function as the last line of the function. This would be too conservative in some cases (function ends with a multi-line string literal, or a close parenthesis on its own line), but it would be pretty close to correct, and for coveralls (as used by Zulip) I think it wouldn't even matter since coveralls considers continuation lines to be uninteresting anyways. Unfortunately it's not totally trivial to do that either; I might add a `default_visit_node` method to `TraverserVisitor` that is called at the start of every method...